bgpd/bgp_vty.c

```diff
@@ -6112,6 +6112,525 @@ ALIAS_HIDDEN(no_neighbor_addpath_tx_bestpath_per_as,
 	     NO_STR NEIGHBOR_STR NEIGHBOR_ADDR_STR2
 	     "Use addpath to advertise the bestpath per each neighboring AS\n")
 
+
+DEFUN_NOSH (vpn_policy_afi,
+       vpn_policy_afi_cmd,
+       "vpn-policy <ipv4|ipv6>",
+       "Enter vpn-policy command mode\n"
+       BGP_AFI_HELP_STR)
+{
+	VTY_DECLVAR_CONTEXT(bgp, bgp);
+	if (bgp->inst_type != BGP_INSTANCE_TYPE_VRF
+	    && bgp->inst_type != BGP_INSTANCE_TYPE_DEFAULT) {
+
+		vty_out(vty,
+			"vpn-policy supported only in core or vrf instances.\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	afi_t afi;
+	int idx = 0;
+
+	if (argv_find_and_parse_afi(argv, argc, &idx, &afi)) {
+		if (afi == AFI_IP)
+			vty->node = BGP_VPNPOLICY_IPV4_NODE;
+		else
+			vty->node = BGP_VPNPOLICY_IPV6_NODE;
+		return CMD_SUCCESS;
+	}
+	return CMD_WARNING_CONFIG_FAILED;
+}
+
+static int vpn_policy_afis(struct vty *vty, int *doafi)
+{
+	switch (vty->node) {
+	case BGP_VPNPOLICY_IPV4_NODE:
+		doafi[AFI_IP] = 1;
+		break;
+	case BGP_VPNPOLICY_IPV6_NODE:
+		doafi[AFI_IP6] = 1;
+		break;
+	default:
+		vty_out(vty,
+			"%% context error: valid only in vpn-policy block\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+	return CMD_SUCCESS;
+}
+
+static int argv_find_and_parse_vpn_policy_dirs(struct vty *vty,
+    struct cmd_token **argv, int argc, int *idx, int *dodir)
+{
+	if (argv_find(argv, argc, "fromvpn", idx)) {
+		dodir[BGP_VPN_POLICY_DIR_FROMVPN] = 1;
+	} else if (argv_find(argv, argc, "tovpn", idx)) {
+		dodir[BGP_VPN_POLICY_DIR_TOVPN] = 1;
+	} else if (argv_find(argv, argc, "both", idx)) {
+		dodir[BGP_VPN_POLICY_DIR_FROMVPN] = 1;
+		dodir[BGP_VPN_POLICY_DIR_TOVPN] = 1;
+	} else {
+		vty_out(vty, "%% direction parse error\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+	return CMD_SUCCESS;
+}
+
+DEFUN (vpn_policy_rd,
+       vpn_policy_rd_cmd,
+       "rd ASN:NN_OR_IP-ADDRESS:NN",
+       "Specify route distinguisher\n"
+       "Route Distinguisher (<as-number>:<number> | <ip-address>:<number>)\n")
+{
+	VTY_DECLVAR_CONTEXT(bgp, bgp);
+	struct prefix_rd prd;
+	int ret;
+	int doafi[AFI_MAX] = {0};
+	afi_t afi;
+
+	ret = str2prefix_rd(argv[1]->arg, &prd);
+	if (!ret) {
+		vty_out(vty, "%% Malformed rd\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	ret = vpn_policy_afis(vty, doafi);
+	if (ret != CMD_SUCCESS)
+		return ret;
+
+
+	for (afi = 0; afi < AFI_MAX; ++afi) {
+		if (!doafi[afi])
+			continue;
+
+		/* pre-change: un-export vpn routes (vpn->vrf routes unaffected)
+		 */
+		vpn_leak_prechange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				   bgp_get_default(), bgp);
+
+		bgp->vpn_policy[afi].tovpn_rd = prd;
+		SET_FLAG(bgp->vpn_policy[afi].flags,
+			 BGP_VPN_POLICY_TOVPN_RD_SET);
+
+		/* post-change: re-export vpn routes */
+		vpn_leak_postchange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				    bgp_get_default(), bgp);
+	}
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (vpn_policy_no_rd,
+       vpn_policy_no_rd_cmd,
+       "no rd",
```

Comment:
Please make this `no rd [ASN:NN_OR_IP-ADDRESS:NN]`. This way users can copy-paste the positive form of the command, prepend `no` and have it negated. Alternatively, you can save code by putting `[no]` at the start of the positive form and handling that.